### PR TITLE
Support Q < D in linear form device kernels

### PR DESCRIPTION
**Pull request by pazner**
Some of the linear form device kernels assume that Q >= D. This PR should fix that issue (at the cost of using somewhat more shared memory) and adds a unit test to check.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#3569](https://github.com/mfem/mfem/pull/3569) | @pazner | @v-dobrev | @kmittal2 + @camierjs | 4/6/23 | 4/28/23 | 5/2/23 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by v-dobrev** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

File: tests/unit/fem/test_linearform_ext.cpp [hunk 1]
```diff
@@ -67,12 +69,23 @@ struct LinearFormExtTest
    LinearForm lf_dev, lf_std;
    LinearFormIntegrator *lfi_dev, *lfi_std;
 
+   static int GetQOrder(QuadratureType qtype, int p)
+   {
+      switch (qtype)
+      {
+         case Lobatto: return 2*p - 1;
+         case Legendre: return 2*p + 3;
+         case LegendreUnderIntegration: return 2*p -1;
```
Comment (camierjs) [inline tests/unit/fem/test_linearform_ext.cpp:78]: `2*p - 1`

File: fem/lininteg_domain_grad.cpp
Comment (camierjs) [inline fem/lininteg_domain_grad.cpp:1]: The kernel could avoid using more shared memory by using `Q` threads when required:

```diff
diff --git a/fem/lininteg_domain_grad.cpp b/fem/lininteg_domain_grad.cpp
index ae8ae3b41..5a64059fa 100644
--- a/fem/lininteg_domain_grad.cpp
+++ b/fem/lininteg_domain_grad.cpp
@@ -126,7 +126,9 @@ void DLFGradAssemble3D(const int vdim, const int ne, const int d, const int q,

    auto Y = Reshape(output, d,d,d, vdim, ne);

-   MFEM_FORALL_2D(e, ne, q, q, 1,
+   static constexpr int Q = T_Q1D ? T_Q1D : MAX_Q1D;
+
+   MFEM_FORALL_2D(e, ne, Q, Q, 1,
    {
       if (M(e) == 0) { return; } // ignore

@@ -137,15 +139,14 @@ void DLFGradAssemble3D(const int vdim, const int ne, const int d, const int q,
       double r_u[D];

       MFEM_SHARED double sBGt[2][Q*D];
-      MFEM_SHARED double sQQQ_1[Q*MQD*MQD];
-      MFEM_SHARED double sQQQ_2[Q*Q*D];
+      MFEM_SHARED double sQQQ[MQD*MQD*MQD];

       const DeviceMatrix Bt(sBGt[0], q,d), Gt(sBGt[1], q,d);
       kernels::internal::LoadBGt<D,Q>(d,q,B,G,sBGt);

-      const DeviceCube QQQ(sQQQ_1, q,q,q);
-      const DeviceCube QQD(sQQQ_2, q,q,d);
-      const DeviceCube QDD(sQQQ_1, q,d,d);
+      const DeviceCube QQQ(sQQQ, q,q,q);
+      const DeviceCube QQD(sQQQ, q,q,d);
+      const DeviceCube QDD(sQQQ, q,d,d);

       for (int c = 0; c < vdim; ++c)
       {
```

File: tests/unit/fem/test_linearform_ext.cpp [hunk 2]
```diff
@@ -67,12 +69,23 @@ struct LinearFormExtTest
...
+      }
+   }
```
[the remark this answers is not in the document]
Comment (pazner) [reply tests/unit/fem/test_linearform_ext.cpp:82]: Fixed in 3581e69476c3338ed28115420a3b088fc4bea278

File: tests/unit/fem/test_linearform_ext.cpp (hunk 1 above)
Comment (pazner) [reply tests/unit/fem/test_linearform_ext.cpp]: Fixed in 3581e69476c3338ed28115420a3b088fc4bea278

File: fem/lininteg_domain_grad.cpp
Comment (pazner) [reply fem/lininteg_domain_grad.cpp:1]: This is much better, thanks! Added in 08a2d97ea14b7a18b2d2dc327e2568bf1f4fc149

**Comment by v-dobrev** [comment]
Note: the last commit has failures.

**Comment by pazner** [comment]
> Note: the last commit has failures.

There was a problem with the version of the kernel committed in 08a2d97, fixed by @camierjs in bfe1fe4.

**Review by camierjs — APPROVED** [review]

**Review by kmittal2 — APPROVED** [review]

**Comment by v-dobrev** [comment]
Merged in `next` for testing...

**Pull request merged**